File: Triangle.java
public class Solution {
	public int minimumTotal(List<List<Integer>> triangle) {
		return minimalTotalRecur(triangle, 0, 0);
	}

	public int minimalTotalRecur(List<List<Integer>> triangle, int row, int position) {
		if(row >= triangle.size() || position >= triangle.get(row).size())
			return 0;
		return triangle.get(row).get(position) + Math.min(minimalTotalRecur(triangle, row + 1, position), minimalTotalRecur(triangle, row + 1, position + 1));
	}
}

File: Triangle.cc
class Solution {
public:
    int minimumTotal(vector<vector<int> > &triangle) {
        vector<vector<int>> dp(triangle.size(), vector<int>());
        for(int i = 0; i < triangle.size(); i++) {
            for(int j = 0; j < triangle[i].size(); j++) {
                dp[i].push_back(triangle[i][j]);
            }
        }

        for(int i = dp.size() - 2; i >= 0; i--) {
            for(int j = 0; j < dp[i].size(); j++) {
                dp[i][j] += min(dp[i+1][j], dp[i+1][j+1]);
            }
        }
        return dp[0][0];
    }
};
